parcels/kernel.py

```diff
@@ -88,119 +39,92 @@ class Kernel(BaseKernel):
-        self.funccode = (  # Remove parcels. prefix (see #1608)
-            self.funccode.replace("parcels.StatusCode", "StatusCode")
-        )
 
-        # Parse AST if it is not provided explicitly
-        self.py_ast = (
-            py_ast or ast.parse(textwrap.dedent(self.funccode)).body[0]
-        )  # Dedent allows for in-lined kernel definitions
-        if pyfunc is None:
-            # Extract user context by inspecting the call stack
-            stack = inspect.stack()
-            try:
-                user_ctx = stack[-1][0].f_globals
-                user_ctx["math"] = globals()["math"]
-                user_ctx["random"] = globals()["random"]
-                user_ctx["StatusCode"] = globals()["StatusCode"]
-            except:
-                warnings.warn(
-                    "Could not access user context when merging kernels",
-                    KernelWarning,
-                    stacklevel=2,
-                )
-                user_ctx = globals()
-            finally:
-                del stack  # Remove cyclic references
-            # Generate Python function from AST
-            py_mod = ast.parse("")
-            py_mod.body = [self.py_ast]
-            exec(compile(py_mod, "<ast>", "exec"), user_ctx)
-            self._pyfunc = user_ctx[self.funcname]
-        else:
-            self._pyfunc = pyfunc
-
-        self.name = f"{ptype.name}{self.funcname}"
+        self._pyfuncs: list[Callable] = pyfuncs
+
+    @property  #! Ported from v3. To be removed in v4? (/find another way to name kernels in output file)
+    def funcname(self):
+        ret = ""
+        for f in self._pyfuncs:
+            ret += f.__name__
+        return ret
+
+    @property  #! Ported from v3. To be removed in v4? (/find another way to name kernels in output file)
+    def name(self):
+        return f"{self._ptype.name}{self.funcname}"
 
     @property
     def ptype(self):
         return self._ptype
 
-    @property
-    def pyfunc(self):
-        return self._pyfunc
-
     @property
     def fieldset(self):
         return self._fieldset
 
+    def remove_deleted(self, pset):
+        """Utility to remove all particles that signalled deletion."""
+        bool_indices = pset._data["state"] == StatusCode.Delete
+        indices = np.where(bool_indices)[0]
+        # TODO v4: need to implement ParticleFile writing of deleted particles
+        # if len(indices) > 0 and self.fieldset.particlefile is not None:
+        #     self.fieldset.particlefile.write(pset, None, indices=indices)
+        if len(indices) > 0:
+            pset.remove_indices(indices)
+
     def add_positionupdate_kernels(self):
         # Adding kernels that set and update the coordinate changes
         def Setcoords(particle, fieldset, time):  # pragma: no cover
             import numpy as np  # noqa
 
-            particle_dlon = 0  # noqa
-            particle_dlat = 0  # noqa
-            particle_ddepth = 0  # noqa
+            particle.dlon = 0
```

Comment:
So now we have three location variables per direction for each particle, increasing the memory footprint of the pset. Is there no way to keep using only two?

Reply:
Can we not do
1. initialise `particle.dlon` = 0
2. At the start of each kennel loop, `particle.lon += particle.dlon`
3. Then run the kernels?

This way, we could get rid of `particle.lon_nextloop`? Or am I missing something?

Reply:
I think that we should do this in a separate PR. I saw that change as being separate, and it would be good to test it properly (i.e., once we have particle file output so that we can compare).

Reply:
OK, fully agree!

Reply:
This effort would then be part of solving #2084